docs/data/data-grid/filtering/filtering.md

```diff
@@ -47,6 +47,41 @@ To disable `Add filter` or `Remove all` buttons, pass `disableAddFilterButton` o
 
 {{"demo": "DisableActionButtonsDataGridPro.js", "bg": "inline", "defaultCodeOpen": false}}
 
+## Header filtering [<span class="plan-pro"></span>](/x/introduction/licensing/#pro-plan)
+
+:::warning
+This feature is experimental, it needs to be explicitly activated using the `headerFiltering` experimental feature flag.
+
+```tsx
+<DataGridPro experimentalFeatures={{ headerFiltering: true }} {...otherProps} />
+```
+
+To use header filtering, you need to upgrade to the [Pro plan](/x/introduction/licensing/#pro-plan) or above.
+
+:::
+
+Header filtering adds a new header row using which the users can quickly add some filters to the columns. The filters added on filter panel are synchronized with the filtering row and vice versa.
+
+You can switch between different operators by clicking the operator button in the header filter cell or pressing <kbd><kbd class="key">Ctrl</kbd>+<kbd class="key">Enter</kbd></kbd> (or <kbd><kbd class="key">⌘ Command</kbd>+<kbd class="key">Enter</kbd></kbd> on macOS) when focusing on a header filter cell.
+
+{{"demo": "HeaderFilteringDataGridPro.js", "bg": "inline", "defaultCodeOpen": false}}
+
+:::info
+**Tip:** You can render custom components in the header filtering row by using the `colDef.renderHeaderFilter` method.
```

Comment:
Tried to do it in https://github.com/mui/mui-x/pull/7760/commits/54e0b17f51bc8d2e358acf964cd7d42a11c150f2 and [updated](https://codesandbox.io/s/3okcng?file=/demo.tsx) the codesandbox to reflect new changes, do you think it makes sense now?

Reply:
> do you think passing the Adorment component with a generic name (let's say headerFilterMenu) and then using that prop inside default Filter components such as GridFilterInputValue etc to make it TextField agnostic and use in custom input components the same way would work?

Yes, I think we should not assume TextField API here.
 
> Tried to do it in https://github.com/mui/mui-x/commit/54e0b17f51bc8d2e358acf964cd7d42a11c150f2 and [updated](https://codesandbox.io/s/3okcng?file=/demo.tsx) the codesandbox to reflect new changes, do you think it makes sense now?

Looks awesome!
One thing I noticed here is that you have to click twice to open the menu if the cell is not focused:

https://user-images.githubusercontent.com/13808724/235452647-753fa66f-4f5c-42e9-96fe-f609185861cd.mov

Reply:
> Another one would be to allow usecases like https://github.com/mui/mui-x/issues/4934#issuecomment-1240320900 (point no. 2) where users would like to use 2nd header row for some reason other than the filtering itself.

I'm unsure about using `renderHeaderFilter` for this, as it seems inefficient - a lot of filter-related computations will be done anyway.
But it will work.

Reply:
> a lot of filter-related computations will be done anyway.

Yes this concern is valid 👍 , one way of solving this that I thought of was to add a slot for the `HeaderFilterItem` and let the users override the whole slot instead of sub-part that is done by `renderHeaderFilter`, but I decided to not go with it for now due to these reasons.
1. It will require the users to receive props like width and height and apply properly for it to work as expected (learning curve)
2. It will be a bit more difficult to do customization.
    - A user want to disable header filter for a specific column.
      1. **Using `renderHeaderFilter`**, they can pass something like `renderHeaderFilter: () => null` in `colDef`
      2. **Using slot:** The user will have to conditionally pass `null` after checking `colDef.field === 'x'` and reproduce the default behavior themselves (copy paste all the default code) which doesn't seem great.
    - Another user want to use a custom header filter for a specific column.
      1. **Using `renderHeaderFilter`**, they can pass something like `renderHeaderFilter: (props) => <CustomHeaderFilter {...props} />` in `colDef`
      2. **Using slot:** The user will have to conditionally pass `<CustomHeaderFilter {...props} />` after checking `colDef.field === 'x'` and again reproduce the default behavior themselves.
    - What if they want to override/customize for all columns?
      Should we introduce something like `slotProps.headerFilter.renderHeaderFilter` which is preceded for the columns where `colDef.renderHeaderFilter` is not passed 🤔 
4. So going with the same pattern like `renderHeader`, `renderCell` and `renderHeaderGroup` seemed more like it.

We can consider point no. 2 mentioned [here](https://github.com/mui/mui-x/pull/7760#discussion_r1179251451) as a corner case which is handled as a bonus along with the basic functionality.

Let me know if you disagree on something as this is a open to discussion topic.